### PR TITLE
ttl: fix the issue that TTL job may hang some time when shrink the delete worker count

### LINKED ISSUE #55561: TTL job hangs in running state after decreasing `tidb_ttl_delete_worker_count`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set @@global.tidb_ttl_scan_batch_size=500;
set @@global.tidb_ttl_delete_batch_size=1;
set @@global.tidb_ttl_delete_rate_limit=1;
set @@global.tidb_ttl_delete_worker_count=2;
create table ttl1(t timestamp) TTL=`t`+interval 1 minute ttl_job_interval='1m';
insert into ttl1 values(now() - interval 1 day);
insert into ttl1 select * from ttl1; -- repeat many times
```

Then wait ttl job is running, after ttl job is running do:

```
set @@global.tidb_ttl_delete_worker_count=1;
```

Wait ttl job finished.

### 2. What did you expect to see? (Required)

The TTL job will finally finish.

### 3. What did you see instead (Required)

The TTL job will never finish because some task is always in running state:

```
> select * from mysql.tidb_ttl_task\G
***************************[ 1. row ]***************************
job_id             | 70a857793b0e4296975fe8f695cb05d5
table_id           | 27866
scan_id            | 0
scan_range_start   |
scan_range_end     |
expire_time        | 2024-08-21 15:01:47
owner_id           | af7d74fd-81d4-4e16-ac78-3a1f4fc6919f
owner_addr         | <null>
owner_hb_time      | 2024-08-21 15:10:08
status             | running
status_update_time | 2024-08-21 15:02:47
state              | {"total_rows":32766,"success_rows":32368,"error_rows":1,"scan_task_err":""}
created_time       | 2024-08-21 15:02:47
1 row in set
```

You can see that `total_rows(32766) > success_rows(32368) + error_rows(1)`. The task will only seen as finished when they are equal.

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #55561

### What changed and how does it work?

- fix some missed count for `ErrorRows`
- if a scan task is finished, it will wait for all rows to be processed. If it is not finished in a certain time, the task will go one to mark itself as finished.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by bb7133 — APPROVED** [review]
LGTM

File: pkg/ttl/ttlworker/task_manager.go [hunk 1]
```diff
@@ -597,6 +601,33 @@ func (t *runningScanTask) Context() context.Context {
...
+		zap.String("jobID", t.JobID),
+		zap.Int64("scanID", t.ScanID),
+		zap.String("table", t.tbl.Name.O),
+	)
+
+	if t.statistics.TotalRows.Load() <= t.statistics.ErrorRows.Load()+t.statistics.SuccessRows.Load() {
+		// All rows are processed.
+		// We use `<=` instead of `==` to make the logic strong to make sure
+		// it also works when statistics are not accurate.
+		logger.Info("mark TTL task finished because all scanned rows are processed")
```
Comment (YangKeao) [inline pkg/ttl/ttlworker/task_manager.go:620]: How about adding a warning for the case `t.statistics.ErrorRows.Load()+t.statistics.SuccessRows.Load() > t.statistics.TotalRows.Load()`, as we currently don't know when they are not equal.

File: pkg/ttl/ttlworker/task_manager.go [hunk 2]
```diff
@@ -597,6 +601,33 @@ func (t *runningScanTask) Context() context.Context {
...
+		return true
+	}
+
+	if time.Since(t.result.time) > waitTaskProcessRowsTimeout {
```
Comment (YangKeao) [inline pkg/ttl/ttlworker/task_manager.go:624]: Is two minutes long enough? Previously, the meaning of `finished()` is that both `scan` and `delete` are finished. If we are going to keep the behavior, it means the `delete` should have finished within 2 minutes, right?

Comment (lcwangchao) [reply pkg/ttl/ttlworker/task_manager.go:624]: Maybe I can enlarge it to 5 minutes. The `delChan` is a channel with 0 zero size, and san worker will only be marked as finished after the `delTask` emitted. That means we as most waiting `tidb_ttl_scan_batch_size` rows to deleted. `tidb_ttl_scan_batch_size` default value is 500, I think it is enough for normal cases

**Review by YangKeao — APPROVED** [review]
LGTM

**Pull request merged**